Fix problem with fromTrackerSeeds parameter in ElectronSeedProducer and improve LayerMeasurements class

guitargeek opened the pull request:
#### PR description:

This PR intends to fix https://github.com/cms-sw/cmssw/issues/27541.

If the `fromTrackerSeeds` parameter in the `ElectronSeedProducer` would be set to `false`, a segfault would occur, as detected in the UBSAN IB. This never happened in practice because  `fromTrackerSeeds = False` is actually not supported and triggers an exception [1].

We understand now that` fromTrackerSeeds` is never supposed to be `False` and therefore we can just remove all branches in the code for this case, fixing both the issue and simplifying the code.

When I tried to fix the issue, I initially didn't know how it could be fixed and did some changes with the goal of making the problem more obvious to me. This involved changing the `LayerMeasurements` class such that there are no pointers anymore which can be `nullptr` and therefore cause trouble when dereferenced. In retrospect, this has nothing to do with the actual issue, but I include this change anyway in the PR as I think it's an improvement.

This caused the interface of `LayerMeasurements` to change a little bit, and while adapting the code I found an unused function in some muon plugins which could be removed. It was apparently replaced by a more efficient function.

[1] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaElectronAlgos/src/ElectronSeedGenerator.cc#L275

#### PR validation:

CMSSW compiles and local matrix tests pass.

#### if this PR is a backport please specify the original PR:

No backports intended.

guitargeek committed:
No pointers anymore in LayerMeasurements class

guitargeek committed:
proper constructor for ElectronSeedGenerator

guitargeek committed:
code format

guitargeek committed:
fixing issue 27541

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28039/11971

- This PR adds an extra 48KB to repository

cmsbuild commented:
A new Pull Request was created by  @guitargeek (Jonas Rembser) for master.

It involves the following packages:

FastSimulation/Muons
RecoEgamma/EgammaElectronAlgos
RecoMuon/TrackerSeedGenerator
RecoTracker/TkTrackingRegions
TrackingTools/MeasurementDet


@perrotta, @ssekmen, @lveldere, @sbein, @civanch, @mdhildreth, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@felicepantaleo, @jainshilpi, @abbiendi, @echapon, @varuns23, @battibass, @makortel, @jhgoh, @lgray, @HuguesBrun, @trocino, @rociovilar, @Sam-Harper, @GiacomoSguazzoni, @rovere, @VinInn, @bellan, @mschrode, @ebrondol, @dgulhan, @calderona, @bachtis, @gpetruc, @matt-komm, @folguera this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: a44148553c68c80e6f660293a9b40ba2858161a9

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a56221/2599/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

guitargeek committed:
correct initianization order in ElectronSeedGenerator

cmsbuild commented:
The code-checks are being triggered in jenkins.

perrotta commented:
please test

perrotta commented:
code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28039/11980

- This PR adds an extra 56KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28039/11980/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28039/11980/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

guitargeek committed:
code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28039/11981

- This PR adds an extra 56KB to repository

cmsbuild commented:
Pull request #28039 was updated. @perrotta, @ssekmen, @lveldere, @sbein, @civanch, @mdhildreth, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 2b4113291aa4a6abb0ce02e58d211b03e0580b7c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a56221/2619/summary.html

cmsbuild commented:
Comparison job queued.

guitargeek committed:
Address comments and some cleanup

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a56221/2619/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2958092
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2957750
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

slava77 commented on RecoEgamma/EgammaElectronAlgos/src/ElectronSeedGenerator.cc:
```diff
@@ -17,188 +17,112 @@
+      phiMax2F_(pset.getParameter<double>("PhiMax2F")),
+      electronMatcher_(pset.getParameter<double>("ePhiMin1"),
+                       pset.getParameter<double>("ePhiMax1"),
+                       phiMin2B_,
```
would std::bind be useful here to be less repetitive in electronMatcher_ and positronMatcher_?

guitargeek replied:
I thought of that too, but concluded it's a waste of time as it is probably only an intermediate solution. There is nothing that makes the electron and positron matcher really different except for a sign flip in the phi-window. IMHO one could just adapt the matcher class a bit and only use one instance after all. I'm afraid doing that would blow up this PR, however.

slava77 commented on RecoEgamma/EgammaElectronProducers/plugins/ElectronSeedProducer.cc:
```diff
@@ -230,16 +230,13 @@ void ElectronSeedProducer::produce(edm::Event& e, const edm::EventSetup& iSetup)
-                                     << ", charge " << (*is).getCharge() << " and cluster energy "
+  for (auto const& seed : *seeds) {
+    SuperClusterRef superCluster = seed.caloCluster().castTo<SuperClusterRef>();
+    LogDebug("ElectronSeedProducer") << "new seed with " << seed.nHits() << " hits"
```
should the loop be placed in ifdef ```EDM_ML_DEBUG``` ? ... that's unless it's obvious that the compiler ignores this loop otherwise.

guitargeek replied:
I think it won't be ignored, I will do as you suggest.

slava77 commented on RecoEgamma/EgammaElectronAlgos/interface/ElectronSeedGenerator.h:
```diff
@@ -38,9 +37,11 @@
 class PropagatorWithMaterial;
 class KFUpdator;
-class PixelHitMatcher;
+class PixelHitMatcher_;
```
is this used in this file?

guitargeek replied:
No, that's an artefact of find-replace. Actually all the forward declarations can be removed.

cmsbuild commented:
The code-checks are being triggered in jenkins.

guitargeek commented:
Hi @slava77, I addressed your comments and further did some trivial cleanup of the `ElectronSeedGenerator` class (removing unused member functions and converting private member functions that don't use the class member variables to free anonymous functions to make this explicit). You are right about the redundancy of the electron and positron matcher, maybe that could be streamlined in another PR eventually.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28039/12081

- This PR adds an extra 56KB to repository

cmsbuild commented:
Pull request #28039 was updated. @perrotta, @ssekmen, @lveldere, @sbein, @civanch, @mdhildreth, @cmsbuild, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

@guitargeek thank you for the quick update

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 20455a4df481ca6acda1d10b83f4b75f9d89e1b6
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a56221/2728/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a56221/2728/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2958092
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2957750
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/28039	20455a4 
- code changes are in line with the PR description and the follow up review; the changes are technical and should not produce any differences in outputs
- jenkins tests pass and comparisons with the baseline show no differences

fabiocos commented:
@ssekmen @sbein could you please have a look? The fastsim part of this PR looks limited

sbein commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
